plugins/zenoh-plugin-storage-manager/src/replica/storage.rs

```diff
@@ -0,0 +1,248 @@
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use crate::storages_mgt::StorageMessage;
+use async_std::sync::Arc;
+use async_std::sync::Mutex;
+use flume::{Receiver, Sender};
+use futures::select;
+use log::{error, trace, warn};
+use std::str;
+use zenoh::key_expr::OwnedKeyExpr;
+use zenoh::prelude::r#async::AsyncResolve;
+use zenoh::prelude::Sample;
+use zenoh::prelude::*;
+use zenoh::time::Timestamp;
+use zenoh::Session;
+use zenoh_backend_traits::{Query, StorageInsertionResult};
+use zenoh_core::Result as ZResult;
+
+pub struct ReplicationService {
+    pub empty_start: bool,
+    pub aligner_updates: Receiver<Sample>,
+    pub log_propagation: Sender<(OwnedKeyExpr, Timestamp)>,
+}
+
+pub struct StorageService {
+    session: Arc<Session>,
+    key_expr: OwnedKeyExpr,
+    name: String,
+    storage: Mutex<Box<dyn zenoh_backend_traits::Storage>>,
+    in_interceptor: Option<Arc<dyn Fn(Sample) -> Sample + Send + Sync>>,
+    out_interceptor: Option<Arc<dyn Fn(Sample) -> Sample + Send + Sync>>,
+    replication: Option<ReplicationService>,
+}
+
+impl StorageService {
+    pub async fn start(
+        session: Arc<Session>,
+        key_expr: OwnedKeyExpr,
+        name: &str,
+        storage: Box<dyn zenoh_backend_traits::Storage>,
+        in_interceptor: Option<Arc<dyn Fn(Sample) -> Sample + Send + Sync>>,
+        out_interceptor: Option<Arc<dyn Fn(Sample) -> Sample + Send + Sync>>,
+        replication: Option<ReplicationService>,
+    ) -> ZResult<Sender<StorageMessage>> {
+        let storage_service = StorageService {
+            session,
+            key_expr,
+            name: name.to_string(),
+            storage: Mutex::new(storage),
+            in_interceptor,
+            out_interceptor,
+            replication,
+        };
+        storage_service.start_storage_queryable_subscriber().await
+    }
+
+    async fn start_storage_queryable_subscriber(&self) -> ZResult<Sender<StorageMessage>> {
+        self.initialize_if_empty().await;
+        let (tx, rx) = flume::bounded(1);
+
+        // subscribe on key_expr
+        let storage_sub = match self
+            .session
+            .declare_subscriber(&self.key_expr)
+            .res_async()
+            .await
+        {
+            Ok(storage_sub) => storage_sub,
+            Err(e) => {
+                error!("Error starting storage {} : {}", self.name, e);
+                return Err(e);
+            }
+        };
+
+        // answer to queries on key_expr
+        let storage_queryable = match self.session.declare_queryable(&self.key_expr).res().await {
+            Ok(storage_queryable) => storage_queryable,
+            Err(e) => {
+                error!("Error starting storage {} : {}", self.name, e);
+                return Err(e);
+            }
+        };
+
+        if self.replication.is_some() {
+            let aligner_updates = &self.replication.as_ref().unwrap().aligner_updates;
+            loop {
```

Comment:
This loop is blocking the caller. It never returns unless stopped via the channel. But the channel's `tx` is returned to the caller only if loop is stopped...
This loop should be running in a side task.

Reply:
Moved the entire `Replica` and `StorageService` into independent tasks. This solves the problem here. Avoiding creating multiple subtasks spawns since it will be problematic/error-prone to manage the replica state then.